File: bin/deploy-check/compare.py
#!/usr/bin/env python3
"""Diff two snapshot.rb captures and report what a deploy changed.

    python3 compare.py before.json after.json

Exit status is 1 when anything in the "must not change" set moved, so this can
gate a deploy in CI. Counts are reported but never fail the run - content changes
while a deploy is in flight.
"""
import json
import sys

# Keys where any change is a regression signal, not normal drift.
CRITICAL = ['available_works', 'consortia', 'profile_path', 'themes',
            'schema_classes', 'features']
# Reported for awareness; expected to move on their own.
INFORMATIONAL = ['counts', 'vocabularies', 'schema_version', 'sample_works',
                 'allowed_by_consortium', 'pending_migrations']


# Intended to move when the submodule is bumped.
EXPECTED_GLOBAL_CHANGES = ['hyrax_version', 'puma_version']


def errored(value):
    """A key the previous version could not even report cannot have regressed."""
    return isinstance(value, str) and value.startswith('ERROR:')


def main(before_path, after_path):
    before = json.load(open(before_path))
    after = json.load(open(after_path))
    problems, notes = [], []

    for key, bval in before.get('global', {}).items():
        aval = after.get('global', {}).get(key)
        if bval == aval:
            continue
        if key in EXPECTED_GLOBAL_CHANGES or errored(bval):
            notes.append(f'global.{key}: {bval!r} -> {aval!r}')
        else:
            problems.append(f'global.{key}: {bval!r} -> {aval!r}')

    btenants, atenants = before.get('tenants', {}), after.get('tenants', {})
    missing = sorted(set(btenants) - set(atenants))
    added = sorted(set(atenants) - set(btenants))
    if missing:
        problems.append(f'tenants disappeared: {missing}')
    if added:
        notes.append(f'tenants added: {added}')

    for name in sorted(set(btenants) & set(atenants)):
        b, a = btenants[name], atenants[name]
        if 'ERROR' in a and 'ERROR' not in b:
            problems.append(f'{name}: now erroring - {a["ERROR"]}')
            continue
        for key in CRITICAL:
            if b.get(key) == a.get(key):
                continue
            if errored(b.get(key)):
                notes.append(f'{name}.{key}: was unreadable before, now {a.get(key)!r}')
            else:
                problems.append(f'{name}.{key}: {b.get(key)!r} -> {a.get(key)!r}')
        for key in INFORMATIONAL:
            if b.get(key) != a.get(key):
                notes.append(f'{name}.{key} changed')
        # A work that had a thumbnail before and lost it is a derivative regression.
        bsamples = {s['id']: s for s in (b.get('sample_works') or []) if isinstance(s, dict)}
        asamples = {s['id']: s for s in (a.get('sample_works') or []) if isinstance(s, dict)}
        for wid, bs in bsamples.items():
            as_ = asamples.get(wid)
            if as_ and bs.get('has_thumbnail') and not as_.get('has_thumbnail'):
                problems.append(f'{name}: work {wid} lost its thumbnail')

    print(f'=== REGRESSIONS ({len(problems)}) ===')
    for p in problems:
        print('  FAIL', p)
    if not problems:
        print('  none - nothing in the must-not-change set moved')
    print(f'\n=== INFORMATIONAL ({len(notes)}) ===')
    for n in notes[:40]:
        print('  note', n)
    if len(notes) > 40:
        print(f'  ... and {len(notes) - 40} more')
    return 1 if problems else 0


if __name__ == '__main__':
    if len(sys.argv) != 3:
        sys.exit(__doc__)
    sys.exit(main(sys.argv[1], sys.argv[2]))

File: bin/deploy-check/compare.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Diff two snapshot.rb captures and report what a deploy changed.
#
#   bin/deploy-check/compare.rb before.json after.json
#
# Exits 1 when anything in CRITICAL moved, so this can gate a deploy. Counts are
# reported but never fail the run, since content changes while a deploy is in flight.
require 'json'

# A change here is a regression signal, not normal drift.
CRITICAL = %w[available_works consortia profile_path themes schema_classes features].freeze
# Reported for awareness; these move on their own.
INFORMATIONAL = %w[counts vocabularies schema_version sample_works
                   allowed_by_consortium pending_migrations].freeze
# Intended to move when the submodule is bumped.
EXPECTED_GLOBAL_CHANGES = %w[hyrax_version puma_version].freeze

# A key the previous version could not even report cannot have regressed.
def errored?(value)
  value.is_a?(String) && value.start_with?('ERROR:')
end

def compare(before, after)
  problems = []
  notes = []

  before.fetch('global', {}).each do |key, bval|
    aval = after.dig('global', key)
    next if bval == aval

    line = "global.#{key}: #{bval.inspect} -> #{aval.inspect}"
    if EXPECTED_GLOBAL_CHANGES.include?(key) || errored?(bval)
      notes << line
    else
      problems << line
    end
  end

  btenants = before.fetch('tenants', {})
  atenants = after.fetch('tenants', {})

  missing = (btenants.keys - atenants.keys).sort
  added = (atenants.keys - btenants.keys).sort
  problems << "tenants disappeared: #{missing.inspect}" if missing.any?
  notes << "tenants added: #{added.inspect}" if added.any?

  (btenants.keys & atenants.keys).sort.each do |name|
    b = btenants[name]
    a = atenants[name]

    if a.key?('ERROR') && !b.key?('ERROR')
      problems << "#{name}: now erroring - #{a['ERROR']}"
      next
    end

    CRITICAL.each do |key|
      next if b[key] == a[key]

      problems << if errored?(b[key])
                    nil
                  else
                    "#{name}.#{key}: #{b[key].inspect} -> #{a[key].inspect}"
                  end
      notes << "#{name}.#{key}: was unreadable before, now #{a[key].inspect}" if errored?(b[key])
    end
    problems.compact!

    INFORMATIONAL.each { |key| notes << "#{name}.#{key} changed" if b[key] != a[key] }

    # A work that had a thumbnail and lost it is a derivative regression.
    bsamples = Array(b['sample_works']).select { |s| s.is_a?(Hash) }.to_h { |s| [s['id'], s] }
    asamples = Array(a['sample_works']).select { |s| s.is_a?(Hash) }.to_h { |s| [s['id'], s] }
    bsamples.each do |id, bs|
      as = asamples[id]
      next unless as && bs['has_thumbnail'] && !as['has_thumbnail']

      problems << "#{name}: work #{id} lost its thumbnail"
    end
  end

  [problems, notes]
end

abort "usage: #{$PROGRAM_NAME} before.json after.json" unless ARGV.length == 2

problems, notes = compare(JSON.parse(File.read(ARGV[0])), JSON.parse(File.read(ARGV[1])))

puts "=== REGRESSIONS (#{problems.size}) ==="
problems.each { |p| puts "  FAIL #{p}" }
puts '  none - nothing in the must-not-change set moved' if problems.empty?

puts "\n=== INFORMATIONAL (#{notes.size}) ==="
notes.first(40).each { |n| puts "  note #{n}" }
puts "  ... and #{notes.size - 40} more" if notes.size > 40

exit(problems.empty? ? 0 : 1)
